llvm18: Binder uses cl.def_readwrite on non const attributes members even if these members do not support operator=

Hello,

Thank you so much for the great work with Binder and pybind11, I have been using it these previous years with success.
I will report 2 issues and suggestions from this experience.

## Issue

Note that this issue happens to me on Ubuntu 24.04 **using llvm18** and did not occur with llvm14.

The first issue I can not get around _very easily_ without being _too intrusive_ occurs when you use **composition** within a **struct A** with a non const **struct B** attribute when B does not support operator= for any of these reasons:

1. B& operator=(const B&); was explicitely deleted
2. B& operator=(const B&); was implicitely deleted (for example when B contains const members)

### Example

#### Input file

```cpp
#pragma once
namespace SampleNamespace {

	class B{
		public:
			B() : i(0) {}
			int const i;
	};

	class A{
		public:
			A() : instanceWithConstMember(B())  {}
			B instanceWithConstMember;
	};
}
```

#### Binder generated file

In this scenario when binding SampleNamespace, Binder generates the following .cpp file:

```cpp
#include <sstream> // __str__

#include <functional>
#include <pybind11/pybind11.h>
#include <string>
#include "limit1.hpp"


#ifndef BINDER_PYBIND11_TYPE_CASTER
	#define BINDER_PYBIND11_TYPE_CASTER
	PYBIND11_DECLARE_HOLDER_TYPE(T, std::shared_ptr<T>, false)
	PYBIND11_DECLARE_HOLDER_TYPE(T, T*, false)
	PYBIND11_MAKE_OPAQUE(std::shared_ptr<void>)
#endif

void bind_unknown_unknown(std::function< pybind11::module &(std::string const &namespace_) > &M)
{
	{ // SampleNamespace::B file: line:4
		pybind11::class_<SampleNamespace::B, std::shared_ptr<SampleNamespace::B>> cl(M("SampleNamespace"), "B", "");
		cl.def( pybind11::init( [](){ return new SampleNamespace::B(); } ) );
		cl.def( pybind11::init( [](SampleNamespace::B const &o){ return new SampleNamespace::B(o); } ) );
		cl.def_readonly("i", &SampleNamespace::B::i);
	}
	{ // SampleNamespace::A file: line:10
		pybind11::class_<SampleNamespace::A, std::shared_ptr<SampleNamespace::A>> cl(M("SampleNamespace"), "A", "");
		cl.def( pybind11::init( [](){ return new SampleNamespace::A(); } ) );
		cl.def_readwrite("instanceWithConstMember", &SampleNamespace::A::instanceWithConstMember);
	}
}
```

Especially:

`cl.def_readwrite("instanceWithConstMember", &SampleNamespace::A::instanceWithConstMember);`

#### Error

I believe that cl.**def_readwrite** implies that _instanceWithConstMember_ could be assigned to some other value of type B, but as B contains const members, the operator= was implicitly deleted from B.

In this context I think that **cl.def_readonly** should have been used instead.

Here is the error I am facing:

```
Compile [...]/unknown/unknown.cpp.o
        In file included from [...]/unknown/unknown.cpp:4:
        /usr/include/pybind11/pybind11.h: In instantiation of 'pybind11::class_<type_, options>& pybind11::class_<type_, options>::def_readwrite(const char*, D C::*, const Extra& ...) [with C = SampleNamespace::A; D = SampleNamespace::B; Extra = {}; type_ = SampleNamespace::A; options = {std::shared_ptr<SampleNamespace::A>}]':
        [...]/unknown/unknown.cpp:27:19:   required from here
        /usr/include/pybind11/pybind11.h:1674:56: error: use of deleted function 'SampleNamespace::B& SampleNamespace::B::operator=(const SampleNamespace::B&)'
         1674 |             fset([pm](type &c, const D &value) { c.*pm = value; }, is_method(*this));
              |                                                  ~~~~~~^~~~~~~
        In file included from [...]/unknown/unknown.cpp:6:
       [...]/limit1.hpp:4:15: note: 'SampleNamespace::B& SampleNamespace::B::operator=(const SampleNamespace::B&)' is implicitly deleted because the default definition would be ill-formed:
            4 |         class B{
              |               ^
        [...]/limit1.hpp:4:15: error: non-static const member 'const int SampleNamespace::B::i', cannot use default assignment operator
```

Replacing generated **def_readwrite** with **def_readonly** prevents this issue from occuring and I verified that the generated library can be used within Python.

## Workaround

The workaround I found was to explicitly implement the **operator= function** for the class, but this often does not make sense, as the const member attributes can not be affected by this function, so I only fill this function with some kind of assert("do not use this function").

## Fix

In case I was missing something like an easier workaround, I did not try to find a fix it in binder's code yet.
Let me know if you consider that this may be fixed in a future release, or if there is a better workaround I did not think of.

Regards,
Ludovic Dedier

--- a/build-and-run-tests.py
+++ b/build-and-run-tests.py
@@ -33,7 +33,7 @@ def main(args):
     parser.add_argument('--binder', default='', help='Path to Binder tool. If none is given then download, build and install binder into main/source/build. Use "--binder-debug" to control which mode of binder (debug/release) is used.')
     parser.add_argument("--binder-debug", action="store_true", help="Run binder tool in debug mode (only relevant if no '--binder' option was specified)")
     parser.add_argument('--pybind11', default='', help='Path to pybind11 source tree')
-    parser.add_argument('--llvm-version', default=None, choices=['6.0.1', '13.0.0', '14.0.5', '18.1.8'], help='Manually specify LLVM version to install')
+    parser.add_argument('--llvm-version', default=None, choices=['6.0.1', '13.0.0', '14.0.5', '18.1.8', '19.1.7'], help='Manually specify LLVM version to install')
     parser.add_argument('--annotate-includes', action="store_true", help='Annotate includes in generated source files')
     parser.add_argument('--trace', action="store_true", help='Binder will add trace output to to generated source files')
     parser.add_argument('--gcc-install-prefix', default=None, help='Path to GCC install prefix which will be used to determent location of libstdc++ for Binder build. Default is: auto-detected. Use this option if you would like to build Binder with compiler that was side-installed and which LLVM build system failed to identify. To see what path Binder uses for libstdc++ run `binder -- -xc++ -E -v`.')

--- a/build-and-run-tests.py
+++ b/build-and-run-tests.py
@@ -33,7 +33,7 @@ def main(args):
     parser.add_argument('--binder', default='', help='Path to Binder tool. If none is given then download, build and install binder into main/source/build. Use "--binder-debug" to control which mode of binder (debug/release) is used.')
     parser.add_argument("--binder-debug", action="store_true", help="Run binder tool in debug mode (only relevant if no '--binder' option was specified)")
     parser.add_argument('--pybind11', default='', help='Path to pybind11 source tree')
-    parser.add_argument('--llvm-version', default=None, choices=['6.0.1', '13.0.0', '14.0.5'], help='Manually specify LLVM version to install')
+    parser.add_argument('--llvm-version', default=None, choices=['6.0.1', '13.0.0', '14.0.5', '18.1.8'], help='Manually specify LLVM version to install')
     parser.add_argument('--annotate-includes', action="store_true", help='Annotate includes in generated source files')
     parser.add_argument('--trace', action="store_true", help='Binder will add trace output to to generated source files')
     parser.add_argument('--gcc-install-prefix', default=None, help='Path to GCC install prefix which will be used to determent location of libstdc++ for Binder build. Default is: auto-detected. Use this option if you would like to build Binder with compiler that was side-installed and which LLVM build system failed to identify. To see what path Binder uses for libstdc++ run `binder -- -xc++ -E -v`.')

--- a/build.py
+++ b/build.py
@@ -165,6 +165,7 @@ def install_llvm_tool(name, source_location, prefix_root, debug, compiler, jobs,
             '13.0.0' : ('https://github.com/llvm/llvm-project/releases/download/llvmorg-13.0.0/llvm-13.0.0.src.tar.xz', 'https://github.com/llvm/llvm-project/releases/download/llvmorg-13.0.0/clang-13.0.0.src.tar.xz', None),
             '14.0.5' : ('https://github.com/llvm/llvm-project/releases/download/llvmorg-14.0.5/llvm-14.0.5.src.tar.xz', 'https://github.com/llvm/llvm-project/releases/download/llvmorg-14.0.5/clang-14.0.5.src.tar.xz', None),
             '18.1.8' : ('https://github.com/llvm/llvm-project/releases/download/llvmorg-18.1.8/llvm-18.1.8.src.tar.xz', 'https://github.com/llvm/llvm-project/releases/download/llvmorg-18.1.8/clang-18.1.8.src.tar.xz', 'https://github.com/llvm/llvm-project/releases/download/llvmorg-18.1.8/cmake-18.1.8.src.tar.xz'),
+            '19.1.7' : ('https://github.com/llvm/llvm-project/releases/download/llvmorg-19.1.7/llvm-19.1.7.src.tar.xz', 'https://github.com/llvm/llvm-project/releases/download/llvmorg-19.1.7/clang-19.1.7.src.tar.xz', 'https://github.com/llvm/llvm-project/releases/download/llvmorg-19.1.7/cmake-19.1.7.src.tar.xz'),
         }[llvm_version]
 
         # The LLVM tarballs may include a "cmake" directory that needs to be
@@ -172,6 +173,7 @@ def install_llvm_tool(name, source_location, prefix_root, debug, compiler, jobs,
         # work. That path can't depend on the LLVM version, so we track that ourselves.
         cmake_path = os.path.join(prefix_root, "cmake")
         cmake_version_path = os.path.join(cmake_path, "llvm_version.txt")
+        hird_party_unittest = os.path.join(prefix_root, 'third-party/unittest')
 
         if not os.path.isfile(prefix + '/CMakeLists.txt') or not os.path.isfile(cmake_version_path) or open(cmake_version_path).read() != llvm_version:
             #execute('Download LLVM source...', 'cd {prefix_root} && curl https://releases.llvm.org/{llvm_version}/llvm-{llvm_version}.src.tar.xz | tar -Jxom && mv llvm-{llvm_version}.src {prefix}'.format(**locals()) )
@@ -190,9 +192,11 @@ def install_llvm_tool(name, source_location, prefix_root, debug, compiler, jobs,
 
         if not os.path.isdir(prefix+'/tools/clang/tools/extra'): os.makedirs(prefix+'/tools/clang/tools/extra')
 
-        if not os.path.isdir(prefix+'/third-party/unittest'):
-            os.makedirs(prefix_root+'/third-party/unittest')
-            with open(prefix_root+'/third-party/unittest/CMakeLists.txt', 'w') as f: f.write('')
+
+        # possible alternative would be to use use '-DBINDER_ENABLE_TEST=OFF', see https://github.com/RosettaCommons/binder/issues/315#issuecomment-2963651447
+        if not os.path.isdir(hird_party_unittest):
+            os.makedirs(hird_party_unittest)
+            with open(hird_party_unittest + '/CMakeLists.txt', 'w') as f: f.write('')
 
 
 
@@ -234,7 +238,7 @@ def install_llvm_tool(name, source_location, prefix_root, debug, compiler, jobs,
         if not os.path.isdir(build_dir): os.makedirs(build_dir)
         execute(
             'Building tool: {}...'.format(name), # -DLLVM_TEMPORARILY_ALLOW_OLD_TOOLCHAIN=1
-            'cd {build_dir} && cmake -G Ninja {config} -DBUILD_TESTING=OFF -DLLVM_ENABLE_EH=1 -DLLVM_ENABLE_RTTI=ON -DLLVM_INCLUDE_BENCHMARKS=OFF {gcc_install_prefix} .. && ninja binder {headers} {jobs}'.format( # was 'binder clang', we need to build Clang so lib/clang/<version>/include is also built
+            'cd {build_dir} && cmake -G Ninja {config} -DLLVM_ENABLE_EH=1 -DLLVM_ENABLE_RTTI=ON -DLLVM_INCLUDE_BENCHMARKS=OFF {gcc_install_prefix} .. && ninja binder {headers} {jobs}'.format( # was 'binder clang', we need to build Clang so lib/clang/<version>/include is also built
                 build_dir=build_dir, config=config,
                 jobs=f'-j{jobs}' if jobs else '',
                 gcc_install_prefix='-DGCC_INSTALL_PREFIX='+gcc_install_prefix if gcc_install_prefix else '',
@@ -284,7 +288,7 @@ def main(args):
     parser.add_argument('--binder', default='', help='Path to Binder tool. If none is given then download, build and install binder into build/ directory. Use "--binder-debug" to control which mode of binder (debug/release) is used.')
     parser.add_argument("--binder-debug", action="store_true", help="Run binder tool in debug mode (only relevant if no '--binder' option was specified)")
     parser.add_argument('--pybind11', default='', help='Path to pybind11 source tree')
-    parser.add_argument('--llvm-version', default=None, choices=['6.0.1', '13.0.0', '14.0.5', '18.1.8'], help='Manually specify LLVM version to install')
+    parser.add_argument('--llvm-version', default=None, choices=['6.0.1', '13.0.0', '14.0.5', '18.1.8', '19.1.7'], help='Manually specify LLVM version to install')
     parser.add_argument('--annotate-includes', action="store_true", help='Annotate includes in generated source files')
     parser.add_argument('--trace', action="store_true", help='Binder will add trace output to to generated source files')
 

--- a/build.py
+++ b/build.py
@@ -131,7 +131,7 @@ def install_llvm_tool(name, source_location, prefix_root, debug, compiler, jobs,
         '14.0.5': 'tools/clang/lib/Headers/clang-resource-headers',
         '13.0.0': 'tools/clang/lib/Headers/clang-resource-headers',
         '6.0.1': 'tools/clang/lib/Headers/clang-headers'
-    }[llvm_version]
+    }.get(llvm_version, 'tools/clang/lib/Headers/clang-resource-headers')
 
     if Platform == 'macos':
         headers += ' clang'
@@ -160,10 +160,11 @@ def install_llvm_tool(name, source_location, prefix_root, debug, compiler, jobs,
 
         clang_path = "{prefix}/tools/clang".format(**locals())
 
-        llvm_url, clang_url = {
-            '6.0.1'  : ('https://releases.llvm.org/6.0.1/llvm-6.0.1.src.tar.xz', 'https://releases.llvm.org/6.0.1/cfe-6.0.1.src.tar.xz'),
-            '13.0.0' : ('https://github.com/llvm/llvm-project/releases/download/llvmorg-13.0.0/llvm-13.0.0.src.tar.xz', 'https://github.com/llvm/llvm-project/releases/download/llvmorg-13.0.0/clang-13.0.0.src.tar.xz'),
-            '14.0.5' : ('https://github.com/llvm/llvm-project/releases/download/llvmorg-14.0.5/llvm-14.0.5.src.tar.xz', 'https://github.com/llvm/llvm-project/releases/download/llvmorg-14.0.5/clang-14.0.5.src.tar.xz')
+        llvm_url, clang_url, cmake_url = {
+            '6.0.1'  : ('https://releases.llvm.org/6.0.1/llvm-6.0.1.src.tar.xz', 'https://releases.llvm.org/6.0.1/cfe-6.0.1.src.tar.xz', None),
+            '13.0.0' : ('https://github.com/llvm/llvm-project/releases/download/llvmorg-13.0.0/llvm-13.0.0.src.tar.xz', 'https://github.com/llvm/llvm-project/releases/download/llvmorg-13.0.0/clang-13.0.0.src.tar.xz', None),
+            '14.0.5' : ('https://github.com/llvm/llvm-project/releases/download/llvmorg-14.0.5/llvm-14.0.5.src.tar.xz', 'https://github.com/llvm/llvm-project/releases/download/llvmorg-14.0.5/clang-14.0.5.src.tar.xz', None),
+            '18.1.8' : ('https://github.com/llvm/llvm-project/releases/download/llvmorg-18.1.8/llvm-18.1.8.src.tar.xz', 'https://github.com/llvm/llvm-project/releases/download/llvmorg-18.1.8/clang-18.1.8.src.tar.xz', 'https://github.com/llvm/llvm-project/releases/download/llvmorg-18.1.8/cmake-18.1.8.src.tar.xz'),
         }[llvm_version]
 
         # The LLVM tarballs may include a "cmake" directory that needs to be
@@ -184,7 +185,15 @@ def install_llvm_tool(name, source_location, prefix_root, debug, compiler, jobs,
             clang_name = 'cfe' if llvm_version == '6.0.1' else 'clang'
             execute('Download Clang source...', 'cd {prefix_root} && curl -LJ {clang_url} | tar -Jxom && mv {clang_name}-{llvm_version}.src {clang_path}'.format(**locals()) )
 
+        if cmake_url:
+            execute('Download CMake source...', 'cd {prefix_root} && curl -LJ {cmake_url} | tar -Jxom && mv cmake-{llvm_version}.src/* {cmake_path}'.format(**locals()) )
+
         if not os.path.isdir(prefix+'/tools/clang/tools/extra'): os.makedirs(prefix+'/tools/clang/tools/extra')
+
+        if not os.path.isdir(prefix+'/third-party/unittest'):
+            os.makedirs(prefix_root+'/third-party/unittest')
+            with open(prefix_root+'/third-party/unittest/CMakeLists.txt', 'w') as f: f.write('')
+
 
 
         # if signature['config'] != disk_signature['config']:
@@ -225,7 +234,7 @@ def install_llvm_tool(name, source_location, prefix_root, debug, compiler, jobs,
         if not os.path.isdir(build_dir): os.makedirs(build_dir)
         execute(
             'Building tool: {}...'.format(name), # -DLLVM_TEMPORARILY_ALLOW_OLD_TOOLCHAIN=1
-            'cd {build_dir} && cmake -G Ninja {config} -DLLVM_ENABLE_EH=1 -DLLVM_ENABLE_RTTI=ON -DLLVM_INCLUDE_BENCHMARKS=OFF {gcc_install_prefix} .. && ninja binder {headers} {jobs}'.format( # was 'binder clang', we need to build Clang so lib/clang/<version>/include is also built
+            'cd {build_dir} && cmake -G Ninja {config} -DBUILD_TESTING=OFF -DLLVM_ENABLE_EH=1 -DLLVM_ENABLE_RTTI=ON -DLLVM_INCLUDE_BENCHMARKS=OFF {gcc_install_prefix} .. && ninja binder {headers} {jobs}'.format( # was 'binder clang', we need to build Clang so lib/clang/<version>/include is also built
                 build_dir=build_dir, config=config,
                 jobs=f'-j{jobs}' if jobs else '',
                 gcc_install_prefix='-DGCC_INSTALL_PREFIX='+gcc_install_prefix if gcc_install_prefix else '',
@@ -275,7 +284,7 @@ def main(args):
     parser.add_argument('--binder', default='', help='Path to Binder tool. If none is given then download, build and install binder into build/ directory. Use "--binder-debug" to control which mode of binder (debug/release) is used.')
     parser.add_argument("--binder-debug", action="store_true", help="Run binder tool in debug mode (only relevant if no '--binder' option was specified)")
     parser.add_argument('--pybind11', default='', help='Path to pybind11 source tree')
-    parser.add_argument('--llvm-version', default=None, choices=['6.0.1', '13.0.0', '14.0.5'], help='Manually specify LLVM version to install')
+    parser.add_argument('--llvm-version', default=None, choices=['6.0.1', '13.0.0', '14.0.5', '18.1.8'], help='Manually specify LLVM version to install')
     parser.add_argument('--annotate-includes', action="store_true", help='Annotate includes in generated source files')
     parser.add_argument('--trace', action="store_true", help='Binder will add trace output to to generated source files')
 

--- a/source/class.cpp
+++ b/source/class.cpp
@@ -124,6 +124,19 @@ bool is_inherited_from_enable_shared_from_this(CXXRecordDecl const *C)
 
 bool is_field_assignable(FieldDecl const *f)
 {
+#if( LLVM_VERSION_MAJOR >= 18 )
+
+	if( CXXRecordDecl *C = f->getType()->getAsCXXRecordDecl() ) { // checking if this type has deleted operator=
+		for( auto m = C->method_begin(); m != C->method_end(); ++m ) {
+			// if( m->getAccess() == AS_public  and  m->isCopyAssignmentOperator()  /*and  !m->doesThisDeclarationHaveABody()*/ and  m->isDeleted() ) return false;
+			if( m->isCopyAssignmentOperator() and (m->getAccess() != AS_public or m->isDeleted()) ) return false;
+		}
+	}
+
+	return true;
+
+#else
+
 	if( RecordType const *r = dyn_cast<RecordType>(f->getType()) ) {
 		if( CXXRecordDecl *C = cast<CXXRecordDecl>(r->getDecl()) ) { // checking if this type has deleted operator=
 			for( auto m = C->method_begin(); m != C->method_end(); ++m ) {
@@ -134,6 +147,7 @@ bool is_field_assignable(FieldDecl const *f)
 	}
 
 	return true;
+#endif
 }
 
 /// check if generator can create binding

--- a/test/T07.class_with_const_members.hpp
+++ b/test/T07.class_with_const_members.hpp
@@ -1,0 +1,34 @@
+
+// -*- mode:c++;tab-width:2;indent-tabs-mode:t;show-trailing-whitespace:t;rm-trailing-spaces:t -*-
+// vi: set ts=2 noet:
+//
+// Copyright (c) 2016 Sergey Lyskov <sergey.lyskov@jhu.edu>
+//
+// All rights reserved. Use of this source code is governed by a
+// MIT license that can be found in the LICENSE file.
+
+/// @author Sergey Lyskov
+/// related issue: https://github.com/RosettaCommons/binder/issues/315
+
+#pragma once
+namespace SampleNamespace {
+
+	class B{
+		public:
+			B() : i(0) {}
+			int const i;
+	};
+
+	class A{
+		public:
+			A() : instanceWithConstMember(B())  {}
+			B instanceWithConstMember;
+	};
+
+	class C{
+		public:
+			C() : muttable(0) {}
+			int muttable;
+	};
+
+}

--- a/test/T07.class_with_const_members.ref.cpp
+++ b/test/T07.class_with_const_members.ref.cpp
@@ -1,0 +1,90 @@
+// File: T07_class_with_const_members.cpp
+#include <T07.class_with_const_members.hpp> // SampleNamespace::A
+#include <T07.class_with_const_members.hpp> // SampleNamespace::B
+#include <T07.class_with_const_members.hpp> // SampleNamespace::C
+#include <sstream> // __str__
+
+#include <functional>
+#include <pybind11/pybind11.h>
+#include <string>
+
+#ifndef BINDER_PYBIND11_TYPE_CASTER
+	#define BINDER_PYBIND11_TYPE_CASTER
+	PYBIND11_DECLARE_HOLDER_TYPE(T, std::shared_ptr<T>, false)
+	PYBIND11_DECLARE_HOLDER_TYPE(T, T*, false)
+	PYBIND11_MAKE_OPAQUE(std::shared_ptr<void>)
+#endif
+
+void bind_T07_class_with_const_members(std::function< pybind11::module &(std::string const &namespace_) > &M)
+{
+	{ // SampleNamespace::B file:T07.class_with_const_members.hpp line:
+		pybind11::class_<SampleNamespace::B, std::shared_ptr<SampleNamespace::B>> cl(M("SampleNamespace"), "B", "");
+		cl.def( pybind11::init( [](){ return new SampleNamespace::B(); } ) );
+		cl.def( pybind11::init( [](SampleNamespace::B const &o){ return new SampleNamespace::B(o); } ) );
+		cl.def_readonly("i", &SampleNamespace::B::i);
+	}
+	{ // SampleNamespace::A file:T07.class_with_const_members.hpp line:
+		pybind11::class_<SampleNamespace::A, std::shared_ptr<SampleNamespace::A>> cl(M("SampleNamespace"), "A", "");
+		cl.def( pybind11::init( [](){ return new SampleNamespace::A(); } ) );
+		cl.def_readonly("instanceWithConstMember", &SampleNamespace::A::instanceWithConstMember);
+	}
+	{ // SampleNamespace::C file:T07.class_with_const_members.hpp line:
+		pybind11::class_<SampleNamespace::C, std::shared_ptr<SampleNamespace::C>> cl(M("SampleNamespace"), "C", "");
+		cl.def( pybind11::init( [](){ return new SampleNamespace::C(); } ) );
+		cl.def_readwrite("muttable", &SampleNamespace::C::muttable);
+	}
+}
+
+
+#include <map>
+#include <algorithm>
+#include <functional>
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+#include <pybind11/pybind11.h>
+
+using ModuleGetter = std::function< pybind11::module & (std::string const &) >;
+
+void bind_T07_class_with_const_members(std::function< pybind11::module &(std::string const &namespace_) > &M);
+
+
+PYBIND11_MODULE(T07_class_with_const_members, root_module) {
+	root_module.doc() = "T07_class_with_const_members module";
+
+	std::map <std::string, pybind11::module> modules;
+	ModuleGetter M = [&](std::string const &namespace_) -> pybind11::module & {
+		auto it = modules.find(namespace_);
+		if( it == modules.end() ) throw std::runtime_error("Attempt to access pybind11::module for namespace " + namespace_ + " before it was created!!!");
+		return it->second;
+	};
+
+	modules[""] = root_module;
+
+	static std::vector<std::string> const reserved_python_words {"nonlocal", "global", };
+
+	auto mangle_namespace_name(
+		[](std::string const &ns) -> std::string {
+			if ( std::find(reserved_python_words.begin(), reserved_python_words.end(), ns) == reserved_python_words.end() ) return ns;
+			return ns+'_';
+		}
+	);
+
+	std::vector< std::pair<std::string, std::string> > sub_modules {
+		{"", "SampleNamespace"},
+	};
+	for(auto &p : sub_modules ) modules[ p.first.empty() ? p.second :  p.first+"::"+p.second ] = modules[p.first].def_submodule( mangle_namespace_name(p.second).c_str(), ("Bindings for " + p.first + "::" + p.second + " namespace").c_str() );
+
+	//pybind11::class_<std::shared_ptr<void>>(M(""), "_encapsulated_data_");
+
+	bind_T07_class_with_const_members(M);
+
+}
+
+// Source list file: TEST/T07_class_with_const_members.sources
+// T07_class_with_const_members.cpp
+// T07_class_with_const_members.cpp
+
+// Modules list file: TEST/T07_class_with_const_members.modules
+// SampleNamespace 

